### PR TITLE
Support deployment to Azure Container Apps with ACA extension

### LINKED ISSUE #3535: Support for deploying to Azure Container Apps
We should add support for deployment to Azure Container Apps. Potentially work with the ACA extension or with shared libraries.

We should also consider having this replace the deployment to ACI.

Depends on https://github.com/microsoft/vscode-azurecontainerapps/issues/209.

### PR DESCRIPTION
**Pull request by bwateratmsft**
Closes #3535, plus a bunch of tracking issues in ADO. Functionality E2E depends on https://github.com/microsoft/vscode-azurecontainerapps/pull/212 / https://github.com/microsoft/vscode-azurecontainerapps/issues/209.

/cc @MicroFish91

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -584,6 +585,11 @@
                     "when": "view == dockerRegistries && viewItem =~ /(DockerV2|DockerHubV2);Tag;/ && isAzureAccountInstalled",
                     "group": "regs_tag_1_general@4"
                 },
+                {
+                    "command": "vscode-docker.registries.deployImageToAca",
+                    "when": "view == dockerRegistries && viewItem =~ /(DockerV2|DockerHubV2);Tag;/ && isAzureAccountInstalled",
```
Comment (bwateratmsft) [inline package.json:590]: Add GitLab if it works

File: src/commands/registries/azure/deployImageToAca.ts [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+import { AzureRegistryTreeItem } from '../../../tree/registries/azure/AzureRegistryTreeItem';
+import { DockerHubNamespaceTreeItem } from '../../../tree/registries/dockerHub/DockerHubNamespaceTreeItem';
+import { DockerV2RegistryTreeItemBase } from '../../../tree/registries/dockerV2/DockerV2RegistryTreeItemBase';
+import { addImageTaggingTelemetry } from '../../images/tagImage';
+
+const acaExtensionId = 'ms-azuretools.vscode-azurecontainerapps';
+const minimumAcaExtensionVersion = '0.4.0'; // TODO: get the exact minimum version that is needed
+
+// The interface of the command options passed to the Azure Container Apps extension's deployImageToAca command
+interface DeployImageToAcaOptionsContract {
```
Comment (bwateratmsft) [inline src/commands/registries/azure/deployImageToAca.ts:23]: Update the param contract with changes from @MicroFish91

File: src/commands/registries/azure/deployImageToAca.ts [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    imageName: string;
+    loginServer?: string;
+    username?: string;
+    secret?: string;
+}
+
+export async function deployImageToAca(context: IActionContext, node?: RemoteTagTreeItem): Promise<void> {
+    // Assert installation of the ACA extension
+    if (!isAcaExtensionInstalled()) {
+        await openAcaInstallPageAndWait(context);
```
Comment (karolz-ms) [inline src/commands/registries/azure/deployImageToAca.ts:33]: Consider just offering to help install the extension, but not attempting to wait for the installation to finish, which is error-prone. It is not a big deal for the user to retry the "deploy image to ACA" gesture.

**Review by karolz-ms — APPROVED** [review]

File: src/commands/registries/azure/deployImageToAca.ts (hunk 3 above)
Comment (bwateratmsft) [reply src/commands/registries/azure/deployImageToAca.ts:33]: Yeah, I think that's reasonable. I wasn't a huge fan of it.

File: src/commands/registries/azure/deployImageToAca.ts [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+import { ext } from '../../../extensionVariables';
+import { registryExpectedContextValues } from '../../../tree/registries/registryContextValues';
+import { RegistryTreeItemBase } from '../../../tree/registries/RegistryTreeItemBase';
+import { AzureRegistryTreeItem } from '../../../tree/registries/azure/AzureRegistryTreeItem';
+import { DockerHubNamespaceTreeItem } from '../../../tree/registries/dockerHub/DockerHubNamespaceTreeItem';
+import { DockerV2RegistryTreeItemBase } from '../../../tree/registries/dockerV2/DockerV2RegistryTreeItemBase';
+import { addImageTaggingTelemetry } from '../../images/tagImage';
+
+const acaExtensionId = 'ms-azuretools.vscode-azurecontainerapps';
+const minimumAcaExtensionVersion = '0.4.0'; // TODO: get the exact minimum version that is needed
```
Comment (bwateratmsft) [inline src/commands/registries/azure/deployImageToAca.ts:20]: Do

File: src/commands/registries/azure/deployImageToAca.ts [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    } else {
+        const { auth } = await registry.getDockerCliCredentials() as { auth?: { username?: string, password?: string } };
+
+        if (!auth?.username || !auth?.password) {
+            throw new Error(localize('vscode-docker.commands.registries.azure.deployImageToAca.noCredentials', 'No credentials found for registry "{0}".', registry.label));
+        }
+
+        if (registry instanceof DockerHubNamespaceTreeItem || registry instanceof DockerV2RegistryTreeItemBase) {
+            // Ensure Docker Hub images are prefixed with 'docker.io/...'
+            if (!/^docker\.io\//i.test(commandOptions.image)) {
```
Comment (bwateratmsft) [inline src/commands/registries/azure/deployImageToAca.ts:61]: This will need a change; it will prepend `docker.io` to non-Docker-Hub images

**Review by rpreslar4765 — DISMISSED** [review]

**Review by karolz-ms — APPROVED** [review]
I guess I need to re-approve this

**Comment by bwateratmsft** [comment]
Associated docs PR: https://github.com/microsoft/vscode-docs/pull/6028

**Pull request merged**